keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/NotificationEventListener.kt

```diff
@@ -0,0 +1,87 @@
+package com.netflix.spinnaker.keel.slack
+
+import com.netflix.spinnaker.keel.api.NotificationType
+import com.netflix.spinnaker.keel.core.api.PromotionStatus
+import com.netflix.spinnaker.keel.events.PinnedNotification
+import com.netflix.spinnaker.keel.events.UnpinnedNotification
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.keel.slack.handlers.SlackNotificationHandler
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.context.event.EventListener
+import org.springframework.stereotype.Component
+import java.time.Clock
+
+/**
+ * Responsible to listen to notification events, and fetch the information needed
+ * for sending a notification, based on NotificationType.
+ */
+@Component
+class NotificationEventListener(
+  private val repository: KeelRepository,
+  private val publisher: ApplicationEventPublisher,
+  private val clock: Clock,
+  private val handlers: List<SlackNotificationHandler<*>>
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  @EventListener(PinnedNotification::class)
+  fun onPinnedNotification(notification: PinnedNotification) {
+    with(notification) {
+      val envNotifications = repository.environmentNotifications(config.name, pin.targetEnvironment)
+      val deliveryArtifact = pin.reference.let {
+        repository.getArtifact(config.name, it)
+      }
+      val pinnedArtifact = repository.getArtifactVersion(deliveryArtifact, pin.version, null)
+      val currentArtifact = repository.getArtifactVersionByPromotionStatus(config, pin.targetEnvironment, deliveryArtifact, PromotionStatus.CURRENT)
+
+      //handlers.supporting<SlackPinnedNotification>(type)
```

Comment:
remove?

Reply:
If you want to try again, this works:
```kotlin
  fun <T : SlackNotificationEvent> Collection<SlackNotificationHandler<*>>.supporting(
    type: com.netflix.spinnaker.keel.notifications.NotificationType
  ): SlackNotificationHandler<T>? =
    handlers.find { hand ->
      hand.type == type
    } as? SlackNotificationHandler<T>
```

Reply:
BTW, we ended up with 2 different `NotificationType` enums, hence the fully-qualified name above. You might want to rename one of them to avoid confusion.